Fusion type checking (#8785)
fishtown-analytics/docs.getdbt.com: website/docs/reference/dbt-jinja-functions/statement-blocks.md

description: "SQL queries that hit database and return results to your Jinja context."
---

<VersionBlock lastVersion="1.99">

:::tip Recommendation

We recommend using the [`run_query` macro](/reference/dbt-jinja-functions/run_query) instead of `statement` blocks. The `run_query` macro provides a more convenient way to run queries and fetch their results by wrapping `statement` blocks. You can use this macro to write more concise code that is easier to maintain.

:::

</VersionBlock>

<VersionBlock firstVersion="2.0">

:::tip When to use statement blocks in Fusion

> For queries where you need to fetch results, you can use either `statement` blocks with `fetch_result=True` or the [`run_query` macro](/reference/dbt-jinja-functions/run_query).

Reviewer: ```suggestion
For queries where you need to fetch results (for example, when your macro or Jinja code needs to use data returned from the database), you can use either `statement` blocks with `fetch_result=True` or the [`run_query` macro](/reference/dbt-jinja-functions/run_query).
```

> For DDL or utility operations (like `OPTIMIZE`, `VACUUM`, or maintenance queries) where you don't need the result set, use `statement` blocks with `fetch_result=False`. This avoids issues with <Constant name="fusion" />'s strict type checking, which can fail when processing result sets that contain null values in columns declared as non-nullable.

Reviewer: ```suggestion
For <Term id="ddl" /> or utility operations (like `OPTIMIZE`, `VACUUM`, or maintenance queries), use `statement` blocks with `fetch_result=False` when you don't need to access the results in Jinja. This avoids issues with <Constant name="fusion" />'s strict type checking, which can fail when processing result sets that contain `NULL` values in columns declared as non-nullable.
```

```

</File>

<VersionBlock firstVersion="2.0">

## Fire and forget operations

> For DDL or utility operations where you don't need the result set, set `fetch_result=False`. This is the recommended pattern for operations like `OPTIMIZE` or `VACUUM` on Databricks, which return result sets that may contain null values in non-nullable columns.

Reviewer: ```suggestion
For <Term id="ddl" /> or utility operations where you don't need the results (because you don't use the returned rows in Jinja), set `fetch_result=False`. This is the recommended pattern for operations like `OPTIMIZE` or `VACUUM` on Databricks, which return result sets that may contain null values in non-nullable columns.
```
